### PR TITLE
Shared cache

### PR DESCRIPTION
**Pull request by CCristi**

### CODE REVIEW

File: src/deep-cache/lib/Driver/CloudFrontDriver.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    });
+  }
+
+  /**
+   * @param {String} key
+   * @param {Object} value
+   * @param {Function} callback
+   * @private
+   */
+  _set(key, value, callback = () => {}) {
```
Comment (devs-deep) [inline src/deep-cache/lib/Driver/CloudFrontDriver.js]: Codacy Issue found: ['callback' is defined but never used.](https://www.codacy.com/app/MitocGroup/deep-framework/file/1585049076/issues/source?bid=2895475&fileBranchId=3024729#l71)

File: src/deep-cache/lib/Driver/CloudFrontDriver.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    throw new MethodNotAvailableException('set');
+  }
+
+  /**
+   * @param key
+   * @param timeout
+   * @param callback
+   * @private
+   */
+  _invalidate(key, timeout = 0, callback = () => {}) {
```
Comment (devs-deep) [inline src/deep-cache/lib/Driver/CloudFrontDriver.js]: Codacy Issue found: ['callback' is defined but never used.](https://www.codacy.com/app/MitocGroup/deep-framework/file/1585049076/issues/source?bid=2895475&fileBranchId=3024729#l81)

File: src/deep-cache/lib/Driver/CloudFrontDriver.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+   * @param {String} url
+   * @param {Function} callback
+   * @private
+   */
+  _request(url, callback) {
+    var client = new XMLHttpRequest();
+
+    client.onreadystatechange = function() {
+      if (client.readyState == 4) {
+        if (client.status != 200) {
```
Comment (devs-deep) [inline src/deep-cache/lib/Driver/CloudFrontDriver.js]: Codacy Issue found: [Expected '!==' and instead saw '!='.](https://www.codacy.com/app/MitocGroup/deep-framework/file/1585049076/issues/source?bid=2895475&fileBranchId=3024729#l104)

File: src/deep-cache/lib/Driver/CloudFrontDriver.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  /**
+   * @param {String} url
+   * @param {Function} callback
+   * @private
+   */
+  _request(url, callback) {
+    var client = new XMLHttpRequest();
+
+    client.onreadystatechange = function() {
+      if (client.readyState == 4) {
```
Comment (devs-deep) [inline src/deep-cache/lib/Driver/CloudFrontDriver.js]: Codacy Issue found: [Expected '===' and instead saw '=='.](https://www.codacy.com/app/MitocGroup/deep-framework/file/1585049076/issues/source?bid=2895475&fileBranchId=3024729#l103)

File: src/deep-cache/lib/Driver/S3FSDriver.js [hunk 5]
```diff
@@ -0,0 +1,118 @@
+/**
+ * Created by CCristi <ccovali@mitocgroup.com> on 12/10/15.
+ */
+
+'use strict';
+
+import {AbstractFsDriver} from './AbstractFsDriver';
+import {Exception} from '../Exception/Exception';
+import path from 'path';
```
Comment (devs-deep) [inline src/deep-cache/lib/Driver/S3FSDriver.js]: Codacy Issue found: ['path' is defined but never used.](https://www.codacy.com/app/MitocGroup/deep-framework/file/1585049085/issues/source?bid=2895475&fileBranchId=3024729#l9)

File: src/deep-cache/lib/Driver/AbstractFsDriver.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+'use strict';
+
+import {AbstractDriver} from './AbstractDriver';
+import crypto from 'crypto';
+
+/**
+ * Abstract FileSystem Driver
+ */
+export class AbstractFsDriver extends AbstractDriver {
+  /**
```
Comment (AlexanderC) [inline src/deep-cache/lib/Driver/AbstractFsDriver.js]: put it at the end of the class

File: src/deep-cache/lib/Driver/AbstractFsDriver.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    this._directory = directory;
+  }
+
+  /**
+   * @param {String} key
+   * @private
+   */
+  _buildKey(key) {
+    return `${this._directory}/${AbstractFsDriver._hash(key, 'sha1')}` +
+           `|${AbstractFsDriver._hash(key, 'md5')}.${this._buildId}`;
```
Comment (AlexanderC) [inline src/deep-cache/lib/Driver/AbstractFsDriver.js]: set a dot "." instead of pipe "|"

Comment (AlexanderC) [inline src/deep-cache/lib/Driver/AbstractFsDriver.js]: add build ID "this._buildId" to the cache file content and invalidate it as it's done for ttl;

File: src/deep-cache/lib/Driver/CloudFrontDriver.js [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    this._request(key, (err, data) => {
+      if (err) {
+        return callback(err, null);
+      }
+
+      try {
+        let parsedData = JSON.parse(data);
+
+        if (parsedData.expires && parsedData.expires < AbstractFsDriver._now) {
+          throw new Exception('Expired');
```
Comment (AlexanderC) [inline src/deep-cache/lib/Driver/CloudFrontDriver.js]: call 'callback(null, null);' and return;

File: src/deep-cache/lib/Driver/S3FSDriver.js [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+        callback(err, null);
+
+        return;
+      }
+
+      try {
+        var parsedData = JSON.parse(data);
+
+        if (parsedData.expires && parsedData.expires < AbstractFsDriver._now) {
+          throw new Exception('Expired');
```
Comment (AlexanderC) [inline src/deep-cache/lib/Driver/S3FSDriver.js]: use

this._invalidate(key);
callback(null, null);
return;

File: src/deep-cache/lib/Driver/S3FSDriver.js [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+   */
+  constructor(containerAware, directory = AbstractFsDriver.DEFAULT_DIRECTORY) {
+    super(directory);
+
+    this._containerAware = containerAware;
+  }
+
+  /**
+   * @param key
+   * @param callback
```
Comment (AlexanderC) [inline src/deep-cache/lib/Driver/S3FSDriver.js]: type hinting?

File: src/deep-cache/lib/Driver/S3FSDriver.js [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+      this._fs.writeFile(key, strObject, (err) => {
+        callback(err, !err);
+      });
+    });
+  }
+
+  /**
+   * @param key
+   * @param timeout
+   * @param callback
```
Comment (AlexanderC) [inline src/deep-cache/lib/Driver/S3FSDriver.js]: type hinting?

File: src/deep-cache/lib/Exception/MethodNotAvailableException.js [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+ * Thrown when a method is not available
+ */
+export class MethodNotAvailableException extends Exception {
+  /**
+   * @param {String} name
+   */
+  constructor(name) {
+    super(`Method '${name}' is not available`);
+  }
+}
```
Comment (AlexanderC) [inline src/deep-cache/lib/Exception/MethodNotAvailableException.js]: newline?

File: src/deep-cache/lib/SharedCache.js [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+    };
+  }
+
+  /**
+   * If key exists do nothing, else create it with specified parameters
+   *
+   * @param key
+   * @param value
+   * @param ttl
+   * @param callback
```
Comment (AlexanderC) [inline src/deep-cache/lib/SharedCache.js]: type hinting?

File: src/deep-cache/lib/SharedCache.js [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+      if (!has) {
+        this._driver.set(key, value, ttl, callback);
+
+        return;
+      }
+
+      callback(null, true);
+    });
+  }
+}
```
Comment (AlexanderC) [inline src/deep-cache/lib/SharedCache.js]: newline?

File: src/deep-cache/lib/Driver/S3FSDriver.js [hunk 15]
```diff
@@ -0,0 +1,122 @@
+/**
+ * Created by CCristi <ccovali@mitocgroup.com> on 12/10/15.
+ */
+
+'use strict';
+
+import {AbstractFsDriver} from './AbstractFsDriver';
+import {Exception} from '../Exception/Exception';
```
Comment (devs-deep) [inline src/deep-cache/lib/Driver/S3FSDriver.js]: Codacy Issue found: ['Exception' is defined but never used.](https://www.codacy.com/app/MitocGroup/deep-framework/file/1597502951/issues/source?bid=2895475&fileBranchId=3024729#l8)

File: src/deep-cache/lib/Driver/CloudFrontDriver.js [hunk 16]
```diff
@@ -0,0 +1,134 @@
+/**
+ * Created by CCristi <ccovali@mitocgroup.com> on 12/10/15.
+ */
+
+'use strict';
+
+import {AbstractFsDriver} from './AbstractFsDriver';
+import {MethodNotAvailableException} from '../Exception/MethodNotAvailableException';
+import {Exception} from '../Exception/Exception';
```
Comment (devs-deep) [inline src/deep-cache/lib/Driver/CloudFrontDriver.js]: Codacy Issue found: ['Exception' is defined but never used.](https://www.codacy.com/app/MitocGroup/deep-framework/file/1597502903/issues/source?bid=2895475&fileBranchId=3024729#l9)

**Pull request merged**